### PR TITLE
[flang] Add ETIME runtime and lowering intrinsics implementation

### PR DESCRIPTION
**Pull request by JumpMasterJJ**
This patch add support of intrinsics GNU extension ETIME  https://github.com/llvm/llvm-project/issues/84205. Some usage info and example has been added to `flang/docs/Intrinsics.md`. The patch contains both the lowering and the runtime code and works on both Windows and Linux.

|   System  |   Implmentation  |
|-----------|--------------------|
| Windows| GetProcessTimes |
| Linux      |times                     |

### CODE REVIEW

File: flang/docs/Intrinsics.md [hunk 1]
```diff
@@ -916,3 +916,54 @@ used in constant expressions have currently no folding support at all.
...
+- **Class:** Subroutine
+- **Syntax:** `CALL ETIME(VALUES, TIME)`
+- **Arguments:**
+
+| Argument   | Description                                                           |
+|------------|-----------------------------------------------------------------------|
+| `VALUES`   | The type shall be REAL(4), DIMENSION(2).                              |
+| `TIME`     | The type shall be REAL(4).                                            |
+
+#### Example
```
Comment (mjklemm) [inline flang/docs/Intrinsics.md:950]: The example is taken from the GFortran website: https://gcc.gnu.org/onlinedocs/gfortran/ETIME.html.  I would suggest to create a different example, or, in the least, refer to the original example on the GFortran website.

File: flang/lib/Evaluate/intrinsics.cpp [hunk 2]
```diff
@@ -1939,7 +1945,7 @@ std::optional<SpecificCall> IntrinsicInterface::Match(
   int elementalRank{0};
   for (std::size_t j{0}; j < dummies; ++j) {
     const IntrinsicDummyArgument &d{dummy[std::min(j, dummyArgPatterns - 1)]};
-    if (const ActualArgument *arg{actualForDummy[j]}) {
+    if (const ActualArgument * arg{actualForDummy[j]}) {
```
Comment (mjklemm) [inline flang/lib/Evaluate/intrinsics.cpp:1948]: These seem to be unrelated formatting changes.  Would you mind taking them back to keep the PR focused on the new feature?  (Same for the other formatting-only changes.)

File: flang/lib/Evaluate/intrinsics.cpp [hunk 3]
```diff
@@ -1340,6 +1340,12 @@ static const IntrinsicInterface intrinsicSubroutine[]{
             {"values", AnyInt, Rank::vector, Optionality::optional,
                 common::Intent::Out}},
         {}, Rank::elemental, IntrinsicClass::impureSubroutine},
+    {"etime",
+        {{"values", TypePattern{RealType, KindCode::exactKind, 4}, Rank::vector,
+             Optionality::required, common::Intent::Out},
+            {"time", TypePattern{RealType, KindCode::exactKind, 4},
+                Rank::scalar, Optionality::required, common::Intent::Out}},
+        {}, Rank::elemental, IntrinsicClass::impureSubroutine},
```
Comment (mjklemm) [inline flang/lib/Evaluate/intrinsics.cpp:1348]: This only supports the `call etime(ta, ti)` pattern of the `ETIME` intrinsic.  However, it also has a function form `ti = etime(at)`.  Support for this would also be needed, as it's being used by applications (despite being discouraged by the GFortran docs).  I was struggling with this, when I tried to make both forms work, as it seems the standard code-gen handlers only support either one.

Comment (klausler) [inline flang/lib/Evaluate/intrinsics.cpp:1348]: `etime` is obviously not elemental.

**Review by klausler — CHANGES_REQUESTED** [review]
`etime` needs to be both a function and a subroutine.

**Review by tblah — COMMENTED** [review]
Welcome to Flang! Thanks for adding this.

File: flang/test/Semantics/etime.f90 [hunk 4]
```diff
@@ -0,0 +1,21 @@
+! RUN: %python %S/test_errors.py %s %flang_fc1 -pedantic
+! Tests for the ETIME intrinsics
+
+subroutine bad_kind_error(values, time)
+  REAL(KIND=8), DIMENSION(2) :: values
+  REAL(KIND=8) :: time
+  !ERROR: Actual argument for 'values=' has bad type or kind 'REAL(8)'
```
[the remark this answers is not in the document]
Comment (mjklemm) [reply flang/test/Semantics/etime.f90:7]: It actually does:

<html>
<body>
<dt style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><em class="emph">Arguments</em>:</dt><dd style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
VALUES | The type shall be REAL(4), DIMENSION(2).
<br/>
TIME | The type shall be REAL(4).

</dd>
</body>
</html>

File: flang/lib/Evaluate/intrinsics.cpp (hunk 3 above)
Comment (mjklemm) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: That's something that had puzzled me too...  All the intrinsic subroutines do it that way, including `cpu_time` and `date_and_time`.  Not sure why that is, but when I started with this I came across it, too, and it got me wondering.

Comment (klausler) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: It's just wrong, and should be fixed.

Comment (mjklemm) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: Ok. But, please, separate PR and fix all of them in one go.  I think this is beyond what @JumpMasterJJ should do in here.

Comment (klausler) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: Are you telling me to go fix them myself?

File: flang/test/Semantics/etime.f90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (tblah) [reply flang/test/Semantics/etime.f90:7]: Ahh sorry about that. Thanks for clarifying

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 5]
```diff
@@ -3230,6 +3235,22 @@ void IntrinsicLibrary::genExecuteCommandLine(
...
+// ETIME
+void IntrinsicLibrary::genEtime(
+    llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 2);
+
+  mlir::Value values = fir::getBase(args[0]);
+  mlir::Value time = fir::getBase(args[1]);
+
+  if (!values)
+    fir::emitFatalError(loc, "expected VALUES parameter");
```
Comment (kkwli) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:3247]: Fortran uses "argument" instead of "parameter".

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 6]
```diff
@@ -3230,6 +3235,22 @@ void IntrinsicLibrary::genExecuteCommandLine(
...
+  if (!time)
+    fir::emitFatalError(loc, "expected TIME parameter");
```
Comment (kkwli) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:3249]: Fortran uses "argument" instead of "parameter".

File: flang/docs/Intrinsics.md (hunk 1 above)
Comment (JumpMasterJJ) [reply flang/docs/Intrinsics.md:950]: Ok, I will add the reference.

File: flang/lib/Evaluate/intrinsics.cpp (hunk 2 above)
Comment (JumpMasterJJ) [reply flang/lib/Evaluate/intrinsics.cpp:1948]: Thank you for your feedback. I'll correct this.

File: flang/runtime/time-intrinsic.cpp [hunk 7]
```diff
@@ -370,5 +374,69 @@ void RTNAME(DateAndTime)(char *date, std::size_t dateChars, char *time,
...
+  double usrTime = -1.0, sysTime = -1.0, realTime = -1.0;
+
+#ifdef _WIN32
+  FILETIME creationTime;
+  FILETIME exitTime;
+  FILETIME kernelTime;
+  FILETIME userTime;
+
+  if (GetProcessTimes(GetCurrentProcess(), &creationTime, &exitTime,
+          &kernelTime, &userTime) != -1) {
```
[the remark this answers is not in the document]
Comment (JumpMasterJJ) [reply flang/runtime/time-intrinsic.cpp:390]: It is indeed a mistake. I'll fix it.

File: flang/runtime/time-intrinsic.cpp [hunk 8]
```diff
@@ -370,5 +374,69 @@ void RTNAME(DateAndTime)(char *date, std::size_t dateChars, char *time,
...
+
+    usrTime = double(userSystemTime.QuadPart / 10000);
+    sysTime = double(kernelSystemTime.QuadPart / 10000);
+    realTime = usrTime + sysTime;
+  }
+#else
+  struct tms tms;
+  if (times(&tms) != -1) {
+    usrTime = (double)(tms.tms_utime) / sysconf(_SC_CLK_TCK);
+    sysTime = (double)(tms.tms_stime) / sysconf(_SC_CLK_TCK);
```
[the remark this answers is not in the document]
Comment (JumpMasterJJ) [reply flang/runtime/time-intrinsic.cpp:405]: It should perform floating division on Windows. I'll update it later.

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp (hunk 5 above)
Comment (JumpMasterJJ) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp:3247]: Ok, but I was referring to the code here. So, should this also be "arguments"?
https://github.com/llvm/llvm-project/blob/410d6350eda322de213941b36adcdab13a0b557b/flang/lib/Optimizer/Builder/IntrinsicCall.cpp#L3193-L3194
https://github.com/llvm/llvm-project/blob/410d6350eda322de213941b36adcdab13a0b557b/flang/lib/Optimizer/Builder/IntrinsicCall.cpp#L3425-L3426

Comment (kkwli) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp:3247]: Yes, I think that needs to be changed as well. Perhaps we can do it in a separate PR.

File: flang/lib/Evaluate/intrinsics.cpp (hunk 3 above)
Comment (mjklemm) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: @klausler Would this acceptable for you?

**Comment by mjklemm** [comment]
> `etime` needs to be both a function and a subroutine.

Yes.  Any recommendation how this can be achieved?  I was trying to add `etime` to the function definitions and the subroutine definitions, but then code-gen handler is not being picked up.

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp (hunk 5 above)
Comment (mjklemm) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp:3245]: Yes, let's keep this PR focused on the `etime` intrinsic.  I will make a PR to clean this up (and also fix up the `elemental` issue with most of the intrinsics, as @klausler as noted correctly).

File: flang/lib/Evaluate/intrinsics.cpp (hunk 3 above)
Comment (klausler) [reply flang/lib/Evaluate/intrinsics.cpp:1348]: Sounds good.

**Comment by klausler** [comment]
> > `etime` needs to be both a function and a subroutine.
> 
> Yes. Any recommendation how this can be achieved? I was trying to add `etime` to the function definitions and the subroutine definitions, but then code-gen handler is not being picked up.

I don't understand that last part.  Does it mean that lowering needs distinct names for the function and the subroutine forms?  That seems likely, and one can do that by defining a distinct name in the function table (say `__builtin_etime_func`) and then defining `etime` as an alias for it in `genericAlias`.

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 9]
```diff
@@ -222,6 +222,11 @@ static constexpr IntrinsicHandler handlers[]{
        {"boundary", asBox, handleDynamicOptional},
        {"dim", asValue}}},
      /*isElemental=*/false},
+    {"etime",
+     &I::genEtime,
+     {{{"values", asBox},
+       {"time", asBox}}},
```
[the remark this answers is not in the document]
Comment (JumpMasterJJ) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp]: I'm not sure about it. I can try it later.

**Comment by mjklemm** [comment]
@JumpMasterJJ Please let me know if you need any help with the pull request to complete the open items.  I can try to help if you need me.

**Comment by JumpMasterJJ** [comment]
> > `etime` needs to be both a function and a subroutine.
> 
> Yes. Any recommendation how this can be achieved? I was trying to add `etime` to the function definitions and the subroutine definitions, but then code-gen handler is not being picked up.

I am trying to declare a `DualGenerator` to solve all intrinsics with both function and subroutine forms.
```C++
// in IntrinsicCall.h

/// Define the different FIR generators that can be mapped to intrinsic to
/// generate the related code.
using ElementalGenerator = decltype(&IntrinsicLibrary::genAbs);
using ExtendedGenerator = decltype(&IntrinsicLibrary::genLenTrim);
using SubroutineGenerator = decltype(&IntrinsicLibrary::genDateAndTime);
// TODO: etime
/// The generator for intrinsic that has both function and subroutine form.
using DualGenerator = decltype(&IntrinsicLibrary::genEtime);
using Generator =
    std::variant<ElementalGenerator, ExtendedGenerator, SubroutineGenerator, DualGenerator>;
```
```C++
// in IntrinsicCall.cpp

// TODO: etime
static fir::ExtendedValue
invokeHandler(IntrinsicLibrary::DualGenerator generator,
              const IntrinsicHandler &handler,
              std::optional<mlir::Type> resultType,
              llvm::ArrayRef<fir::ExtendedValue> args, bool outline,
              IntrinsicLibrary &lib) {
  if (handler.isElemental)
    return lib.genElementalCall(generator, handler.name, mlir::Type{}, args,
                                outline);
  if (outline)
    return lib.outlineInExtendedWrapper(generator, handler.name, resultType,
                                        args);

  if (resultType.has_value()) {
    // function form
    return std::invoke(generator, lib, *resultType, args);
  } else {
    // subroutine form
    std::invoke(generator, lib, args);
  }
  return mlir::Value{};
}
```

How about this plan? Do you have any comments

**Comment by JumpMasterJJ** [comment]
> @JumpMasterJJ Please let me know if you need any help with the pull request to complete the open items. I can try to help if you need me.



> Please let me know if you need any help with the pull request to complete the open items. I can try to help if you need me.

Thank you for your offer. I'll ask for your help if I need more assistance. It's great to know you're there to help.

**Comment by mjklemm** [comment]
> How about this plan? Do you have any comments

I think introducing a new category will be fine.  I'm actually not sure how many of these intrinsics exist (I haven't checked the GFortran docs).  So, maybe the alias route might be the easier one for this instance.

**Comment by JumpMasterJJ** [comment]
> I think introducing a new category will be fine. I'm actually not sure how many of these intrinsics exist (I haven't checked the GFortran docs). So, maybe the alias route might be the easier one for this instance.

Actually, Gfortran has a total of 25 such intrinsics. Here are a few examples:
https://gcc.gnu.org/onlinedocs/gfortran/CHDIR.html
https://gcc.gnu.org/onlinedocs/gfortran/FGET.html
https://gcc.gnu.org/onlinedocs/gfortran/HOSTNM.html
...

**Comment by JumpMasterJJ** [comment]
@mjklemm I just updated the function form of etime, is this acceptable to you?

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 10]
```diff
@@ -1677,6 +1683,24 @@ IntrinsicLibrary::genElementalCall<IntrinsicLibrary::SubroutineGenerator>(
   return mlir::Value();
 }
 
+template <>
+fir::ExtendedValue
+IntrinsicLibrary::genElementalCall<IntrinsicLibrary::DualGenerator>(
```
Comment (mjklemm) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:1688]:
```suggestion
IntrinsicLibrary::genElementalCall<IntrinsicLibrary::DualGenerator> (
```

**Review by mjklemm — APPROVED** [review]
Thanks for updating the PR to also cover the function form of `etime`.  Except for my very tiny nit, I think the PR looks good.

File: flang/lib/Evaluate/intrinsics.cpp [hunk 11]
```diff
@@ -973,6 +977,12 @@ static const std::pair<const char *, const char *> genericAlias[]{
     {"__builtin_ieee_selected_real_kind", "selected_real_kind"},
 };
 
+// Collection for some intrinsics with function and subroutine form,
+// in order to pass
```
Comment (klausler) [inline flang/lib/Evaluate/intrinsics.cpp:981]: Comment appears to have been truncated.

File: flang/lib/Evaluate/intrinsics.cpp [hunk 12]
```diff
@@ -2543,6 +2560,13 @@ bool IntrinsicProcTable::Implementation::IsIntrinsic(
     const std::string &name) const {
   return IsIntrinsicFunction(name) || IsIntrinsicSubroutine(name);
 }
+bool IntrinsicProcTable::Implementation::IsDualIntrinsic(
```
Comment (klausler) [inline flang/lib/Evaluate/intrinsics.cpp:2563]: The table of names could be local to this function.

**Comment by joker-eph** [comment]
This broke this bot: https://lab.llvm.org/buildbot/#/builders/272/builds/16718

I reverted in: https://github.com/llvm/llvm-project/pull/92354
